proposals/XXXX-ConcurrentValueAndConcurrentClosures.md

```diff
@@ -0,0 +1,720 @@
+2. We want advanced programmers to be able to implement libraries with sophisticated techniques (e.g. a concurrent hash table) that can be used in a safe way by others.
+3. We need to embrace the existing world, which contains a lot of code that wasn’t designed with the Swift Concurrency model in mind. We need a smooth and incremental migration story.
+
+Before we jump into the proposed solution, let’s take a look at some common cases that we would like to be able to model along with the opportunities and challenges of each. This will help us reason about the design space we need to cover.
+
+
+### 💖 Swift + Value Semantics
+
+The first kind of type we need to support are simple values like integers. These can be trivially passed across concurrency domains because they do not contain pointers.
+
+Going beyond this, Swift has a strong emphasis on types with [value semantics](https://en.wikipedia.org/wiki/Value_semantics), which are safe to transfer across concurrent boundaries. Except for classes, Swift’s mechanisms for type composition provide value semantics when their elements do. This includes generic structs, as well as its core collections: for example, `Dictionary&lt;Int, String>` can be directly shared across concurrency domains. Swift’s Copy on Write approach means that collections can be transferred without proactive data copying of their representations -- an extremely powerful fact that I believe will make the Swift concurrency model more efficient than other systems in practice.
+
+However, everything isn’t simple here: the core collections can **not** be safely transferred across concurrency domains when they contain general class references, closures that capture mutable state, and other non-value types. We need a way to differentiate between the cases that are safe to transfer and those that are not.
+
+
+### Value Semantic Composition
+
+Structs, enums and tuples are the primary mode for composition of values in Swift. These are all safe to transfer across concurrency domain -- so long as the data they contain is itself safe to transfer.
+
+
+### Higher Order Functional Programming
+
+It is common in Swift and other languages with functional programming roots to use [higher-order programming](https://en.wikipedia.org/wiki/Higher-order_function), where you pass functions to other functions.  Functions in Swift are reference types, but many functions are perfectly safe to pass across concurrency domains - for example, those with an empty capture list.
+
+There are many useful reasons why you’d want to send bits of computation between concurrency domains in the form of a function - even trivial algorithms like `parallelMap` need this.  This occurs at larger scale as well -- for example, consider an actor example like this:
+
+
+```
+actor MyContactList { 
+  func filteredElements(_ fn: (ContactElement) -> Bool) async
+     -> [ContactElement] { … }
+}
+```
```

Comment:
Could this example use [`AsyncSequence`](https://github.com/apple/swift-evolution/blob/main/proposals/0298-asyncsequence.md) and its `filter(_:)` method?

```swift
actor MyContactList: AsyncSequence { ... }
```

In which case, would the `@concurrent` attribute be added to `filter(_:)` in the protocol itself, or in each of its conformances?

Reply:
I think it is best to keep actors conforming to protocols out of this proposal, just because it isn't nailed down yet.